File: android/src/main/kotlin/com/emarsys/emarsys_sdk/command/predict/TrackCartItemCommand.kt
package com.emarsys.emarsys_sdk.command.predict

import com.emarsys.Emarsys
import com.emarsys.emarsys_sdk.command.EmarsysCommand
import com.emarsys.emarsys_sdk.command.ResultCallback
import com.emarsys.predict.api.model.PredictCartItem

class TrackCartItemCommand : EmarsysCommand {
    override fun execute(parameters: Map<String, Any?>?, resultCallback: ResultCallback) {
        val items: List<Map<String, Any>>? = parameters?.get("items") as List<Map<String, Any>>?
        items ?: throw IllegalArgumentException("items should not be null!")

        Emarsys.predict.trackCart(
            items.map {
                PredictCartItem(
                    it["itemId"] as String,
                    it["price"] as Double,
                    it["quantity"] as Double
                )
            })

        resultCallback.invoke(null, null)
    }


    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        return true
    }

    override fun hashCode(): Int {
        return javaClass.hashCode()
    }
}

File: ios/Classes/Wrapper/Commands/Predict/TrackCartItemCommand.swift
//
//  Created by Emarsys
//

import EmarsysSDK

class TrackCartItemCommand: EmarsysCommandProtocol {
    func execute(arguments: [String: Any]?, resultCallback: @escaping ResultCallback) {
        guard let items = arguments?["items"] as? [[String:Any]] else {
            resultCallback(["error": "Invalid items argument"])
            return
        }
        let cartItems:[EMSCartItem] = items.map { cartItem in
            EMSCartItem(
                itemId: cartItem["itemId"] as? String,
                price: cartItem["price"] as! Double,
                quantity: cartItem["quantity"] as! Double
            )
        }
  
        Emarsys.predict.trackCart(items: cartItems)

        resultCallback(["success": true])
    }
}
